### ISSUE TITLE
Enquiry on Trilinear Kernel

### ISSUE THREAD

**Issue by yanxiang-wang** [uzh-rpg/rpg_event_representation_learning#8]
Dear author,
I am interested in your paper and I do not understand what is `trilinear kernel` is.
I found the implantation from `models.py` where there is one step to `init` the kernel.
```python
def trilinear_kernel(self, ts, num_channels):
        gt_values = torch.zeros_like(ts)

        gt_values[ts > 0] = (1 - (num_channels-1) * ts)[ts > 0]
        gt_values[ts < 0] = ((num_channels-1) * ts + 1)[ts < 0]

        gt_values[ts < -1.0 / (num_channels-1)] = 0
        gt_values[ts > 1.0 / (num_channels-1)] = 0

        return gt_values
```
However, I do not figure out what the meaning of the above code. It should be the `gt` values of one `trilinear kernel`.
Then, I write one demo to test it. As follows:
```python
﻿﻿import torch

def trilinear_kernel(ts, num_channels):
    gt_values = torch.zeros_like(ts)

    gt_values[ts > 0] = (1 - (num_channels - 1) * ts)[ts > 0]
    gt_values[ts < 0] = ((num_channels - 1) * ts + 1)[ts < 0]

    gt_values[ts < -1.0 / (num_channels - 1)] = 0
    gt_values[ts > 1.0 / (num_channels - 1)] = 0

    return gt_values

ts = torch.rand((1, 2000))
ts.uniform_(-1, 1)
print(trilinear_kernel(ts,9))
```
The problem is the result is `0` no matter how much times I run.

**Comment by Jee-King** [comment]
Hi, I found that the provided code did not use the **trilinear kernel** or **trilinear_init.pth**(just load it), but used the linear layer three times. Is using three linear layers as the **trilinear kernel**?

**Comment by danielgehrig18** [comment]
Hi Jeeking,

The trilinear kernel in this case is just atriangular kernel applied to t
(so not tri but 1). However, it can be generalized to apply also to x and y
but was not considered in this work.

**Comment by Jee-King** [comment]
Thank you for your prompt reply.

**Comment by Jee-King** [comment]
if we want to apply to  x and y, 
_values = t * self.value_layer.forward(t-i_bin/(C-1))_
we just need to use **x** and **y** instead of **t** in this code. right？

**Comment by danielgehrig18** [comment]
Yes exactly

**Comment by yanxiang-wang** [comment]
Thus, what is the meaning of `gt_values`. Is there one formula to define it?

**Comment by Jee-King** [comment]
the `gt_values `are not all 0.

`print(torch.max(trilinear_kernel(ts,9)))`

output:` tensor(0.9907)`

**Comment by yanxiang-wang** [comment]
> the `gt_values `are not all 0.
> 
> `print(torch.max(trilinear_kernel(ts,9)))`
> 
> output:` tensor(0.9907)`

Yes, I found that. But I cannot understand how to get the `gt_values` since they seem to be calculated from the fixed equations. And the channel number is one parameter for this.
```
gt_values[ts > 0] = (1 - (num_channels - 1) * ts)[ts > 0]
gt_values[ts < 0] = ((num_channels - 1) * ts + 1)[ts < 0]

gt_values[ts < -1.0 / (num_channels - 1)] = 0
gt_values[ts > 1.0 / (num_channels - 1)] = 0
```

**Comment by Jee-King** [comment]
@yanxiang-wang  方便留微信或者QQ吗，有些问题想请教 :blush:
